File: Longest Common Subsequence/longest common subsequence.py
{\rtf1\ansi\ansicpg1252\cocoartf1404\cocoasubrtf470
{\fonttbl\f0\fswiss\fcharset0 Helvetica;}
{\colortbl;\red255\green255\blue255;}
\paperw11900\paperh16840\margl1440\margr1440\vieww10800\viewh8400\viewkind0
\pard\tx566\tx1133\tx1700\tx2267\tx2834\tx3401\tx3968\tx4535\tx5102\tx5669\tx6236\tx6803\pardirnatural\partightenfactor0

\f0\fs24 \cf0 class Solution:\
    def longestCommonSubsequence(self, text1: str, text2: str) -> int:\
        self.dp = \{ \}\
        self.text1 = text1\
        self.text2 = text2\
        n = len(text1)\
        m = len(text2)\
        \
        return self.lcs(n - 1, m - 1)\
        \
    def lcs(self, i, j):\
        if (i == -1 or j == -1):\
            return 0\
        else:\
            if ((i, j) not in self.dp):\
                if (self.text1[i] == self.text2[j]):\
                    self.dp[(i, j)] = self.lcs(i - 1, j - 1) + 1\
                else:\
                    self.dp[(i, j)] = max(self.lcs(i-1, j), self.lcs(i, j-1))\
            return self.dp[(i, j)]}

File: Longest Common Subsequence/longest common subsequence.java
class Solution {
    private String text1, text2;
    private int[][] dp;
    public int longestCommonSubsequence(String text1, String text2) {
        this.text1 = text1;
        this.text2 = text2;
        int n = text1.length();
        int m = text2.length();
        
        this.dp = new int[n][m];
        return lcs(n-1, m-1);
    }
    
    int lcs(int i, int j) {
        if (i == -1 || j == -1) {
            return 0;
        }
        else {
            if (this.dp[i][j] != 0) {
                return this.dp[i][j];
            }
            
            if (text1.charAt(i) == text2.charAt(j)) {
                this.dp[i][j] = lcs(i-1, j-1) + 1;
            }
            else {
                this.dp[i][j] = Math.max(lcs(i-1, j), lcs(i, j-1));
            }
            return this.dp[i][j];
        }
    }
}